### PR TITLE
Add implementation of methods for F_p^n consistent with new API

### PR DESCRIPTION
**Pull request by alexander-zw**
Also changed file structure for fields

### CODE REVIEW

File: libff/algebra/field_utils/bigint.tcc [hunk 1]
```diff
@@ -60,6 +62,42 @@ bigint<n>::bigint(const mpz_t r) /// Initialize from MPZ element
     mpz_clear(k);
 }
 
+template<mp_size_t n>
+bigint<n> bigint<n>::operator-(const bigint<n> &other) const
```
Comment (ValarDragon) [inline libff/algebra/field_utils/bigint.tcc:75]: I'm pretty confused about why this changed. Did most of the functions become non-mutative?

**Comment by ValarDragon** [comment]
I want to check how performance has changed, and if its pretty similar to prior to this PR, it looks good to me! Nice work!

File: libff/algebra/field_utils/bigint.tcc (hunk 1 above)
Comment (alexander-zw) [reply libff/algebra/field_utils/bigint.tcc:75]: These functions are needed to implement Tonelli-Shanks parameter creation, but if those parameters are hard-coded they shouldn't be needed.

**Review by alexander-zw — COMMENTED** [review]
> I want to check how performance has changed, and if its pretty similar to prior to this PR, it looks good to me! Nice work!

I haven't really touched any of the old functions though, just added new ones. Why would performance change?

File: libff/algebra/field_utils/bigint.tcc (hunk 1 above)
Comment (ValarDragon) [reply libff/algebra/field_utils/bigint.tcc:75]: Does Tonelli-Shanks need big integer arithmetic (as opposed to field arithmetic)?

Comment (alexander-zw) [reply libff/algebra/field_utils/bigint.tcc:75]: Tonelli-Shanks itself does not, but to calculate the parameters you need for it you do. We'll usually just hard-code those parameters though, so these functions won't really be needed.

Comment (ValarDragon) [reply libff/algebra/field_utils/bigint.tcc:75]: If so, lets add some comments to indicate that these aren't used in the high performance field arithmetic code. (Still feels a bit weird to have these)

Comment (ValarDragon) [reply libff/algebra/field_utils/bigint.tcc:75]: race condition in comments!

Ok that makes sense, lets add some comments for that. (That way people reviewing field arithmetic know that this isn't done in runtime)

Comment (ValarDragon) [reply libff/algebra/field_utils/bigint.tcc:75]: Per slack discussion, we are removing this code, as it should be generated from Sage. 

Feel free to keep it around in a separate branch for potentially one day defining fields in run time.

**Review by ValarDragon — APPROVED** [review]
LGTM!

**Comment by ValarDragon** [comment]
> I haven't really touched any of the old functions though, just added new ones. Why would performance change?

Now I agree. This wasn't clear to me earlier due to the large code diff from the bigint arithmetic

**Pull request merged**